File: core/internet/source_lineage_graph_engine.py
from __future__ import annotations

from typing import Any, Dict

from core.internet.semantic_provenance_engine import build_semantic_provenance


def build_source_lineage_graph(text: str, url: str = "") -> Dict[str, Any]:
    prov = build_semantic_provenance(text, url)
    citations = prov.get("citations", []) if isinstance(prov.get("citations"), list) else []
    nodes = [url] + [str(c)[:40] for c in citations[:30]]
    edges = [{"from": nodes[i], "to": nodes[i + 1], "relation": "cites"} for i in range(max(0, len(nodes) - 1))]
    return {"nodes": nodes, "edges": edges, "lineage": prov.get("lineage", {})}

File: src/internet/sourceLineageGraphEngine.ts
/**
 * Converted from Python: core/internet/source_lineage_graph_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { buildSemanticProvenance } from "./semanticProvenanceEngine.js";

export function buildSourceLineageGraph(text: any, url: any = ""): any {
  var prov: any = buildSemanticProvenance(text, url);
  var citations: any = ((Array.isArray(py.get(prov, "citations"))) ? py.get(prov, "citations", []) : []);
  var nodes: any = py.add([url], py.iter(py.slice(citations, null, 30)).map((c: any) => py.slice(py.toStr(c), null, 40)));
  var edges: any = py.range(py.max([0, py.sub(py.len(nodes), 1)])).map((i: any) => ({"from": py.at(nodes, i), "to": py.at(nodes, py.add(i, 1)), "relation": "cites"}));
  return {"nodes": nodes, "edges": edges, "lineage": py.get(prov, "lineage", {})};
}
export { buildSemanticProvenance };
